lib/mixlib/shellout.rb

```diff
@@ -199,9 +202,48 @@ def uid
 
     # The gid that the subprocess will switch to. If the group attribute is
     # given as a group name, it is converted to a gid by Etc.getgrnam
+    # If a group is not provided but with_logon is true, the user's primary
+    # group will be used.
     def gid
-      return nil unless group
-      group.kind_of?(Integer) ? group : Etc.getgrnam(group.to_s).gid
+      return group.kind_of?(Integer) ? group : Etc.getgrnam(group.to_s).gid if group 
+      return Etc.getpwuid(uid).gid if (with_logon && user)
```

Comment:
`with_login && user` is repeated a lot.  since you raise if user is not set in conjunction with with_login you don't need to test this at all, or you could move it out to a method `#doing_logon?` or something.